### ISSUE TITLE
Training trick ...

### ISSUE THREAD

**Issue by VanitarNordic** [AlexeyAB/darknet#31]
Hello,

is it good that we stop straining after 1000 iterations and continue training with `yolo-obj_1000.weights`?

or is it good that we go from the beginning to end with `darknet19_448.conv.23`?

My personal assumption is the first one, I mean interruption at 1000 (or maybe sooner) and continue with the new weights, which I believe it will deliver higher accuracy/mAP, but you please confirm the best.

I asked this in a new topic because it was different and it could be helpful for others as well. Thanks.

**Comment by AlexeyAB** [comment]
What is the advantage?

--------

1. The general rule for deep-learning is - The more training data, the better.

2. `darknet19_448.conv.23` trained on the ImageNet 1000-class competition dataset with **~1.4 millions of very carefully labaled images**.

3. Even, if you will train 1000 iterations more (with your 8000 images), then `yolo-obj_1000.weights` will be trained a total of  1 408 000 images. **This is only 0.5% more**. 

But the usefulness of each new piece of data is gradually decreasing. At best, you get an improvement of 0.1%. 

--------

This trick can has an advantage only:

* if you will train ~10000 iterations on large and good image-dataset with a narrow category of images, for example, mechanisms (cars, airs, factory)

* then will extract convolutional layers similar to `./darknet partial cfg/yolo_obj.cfg yolo-obj_10000.weights yolo-obj_10000.conv.23 23`

* and you will use `yolo-obj_10000.conv.23` as pre-trained model to train your new model with the same category of objects - **mechanisms (cars, airs, factory)**: `./darknet detector train data/obj.data yolo-obj.cfg yolo-obj_10000.conv.23`

**Comment by VanitarNordic** [comment]
No, let me explain what I meant.
Assume that we want to detect for example "cars". we start training with `darknet19_448.conv.23` of course. My point is here, should we continue till the end of training with this or should we stop at 1000 iterations for example and continue training with `yolo-voc_1000.weights`?

**Comment by AlexeyAB** [comment]
An identical result. It is the same.

**Comment by VanitarNordic** [comment]
Alright, you know I asked this because I was getting less errors when I was interrupting the training at 1000.

**Comment by AlexeyAB** [comment]
You can see the differences every time when you do the same, because the images for training are selected randomly: https://github.com/AlexeyAB/darknet/blob/b831db5e58ad37c8f4adf39de7fb5db204209a1c/src/data.c#L49

1. How many times did you try to train by using `darknet19_448.conv.23` and train with stoped at 1000 iterations and continue? 

2. How many iterations were in both cases finally?

3. What was the value of the mAP and Recall in both cases for each trial?

This trick can be used only for multi-GPU training: to train 1000 iterations on 1xGPU / stop/ and continue on 4xGPU. This gives the network time to adjust its filters to work better with your current dataset. Then after 1000 iterations many networks (one for each GPU) can be synchronized easily.

**Comment by VanitarNordic** [comment]
1) I did that twice, both cases avg error was lower.
2) I continued to 2000 and 3000 (one class).
3) I don't know about the mAP because it does not printed, I was checking errors and recalls. I can say recalls were identical (with interruption or without interruption) but errors were less with interruption at 1000.

**Comment by AlexeyAB** [comment]
Did you use this fork? https://github.com/VanitarNordic/darknet

**Comment by VanitarNordic** [comment]
I used your latest repository. I downloaded a week ago.

**Comment by VanitarNordic** [comment]
I used your latest repository. I downloaded a week ago.
https://github.com/AlexeyAB/darknet

**Comment by AlexeyAB** [comment]
It's very strange, perform more experiments (~5-10), provide your cfg-file, and write the exact values of iterations and number of images in dataset, and: RPs/Img, IOU, Recall, by using: `darknet.exe detector recall data/obj.data yolo_obj.cfg yolo_obj_3000.weights`

It may be some bug.

**Comment by VanitarNordic** [comment]
Okay. I will do that for sure and I'll post the results here with photos. You did not mention where can I check the mAP which you referred to? if we had that value, then investigation was easier.

I have another question waiting for your reply at previous issue.

**Comment by VanitarNordic** [comment]
Alright, the issue of `544*544` resolution has been solved. 
The problem, as we both guessed, was training image sizes. The training image sizes must not be smaller than the defined network size.
On GTX1060 6G, one images was processed at 41~48ms. it means in case of a video input, processing would be around 20fps or more. The model was trained to detect one class. Again, I amazed that images sizes do not affect the consumed GPU memory. Many of training images were above 2000*2000 in resolution. This is unique for YOLO. Fantastic.

I'll provide test results for training interruption at 1000 iterations anytime soon.

**Comment by VanitarNordic** [comment]
Okay, I had promised to provide some evidences of interrupting effect. 

I mean training interruption in the middle of training, or go with `darknet19_448.conv.23` weights till the end. It seems training interruption caused significant lower avg errors!

Please have a look at these two pictures. Both show 2000+ iterations. One interrupted and restarted and another is the continuation from the beginning:

**1) No interrupting:**

![2000](https://cloud.githubusercontent.com/assets/18719591/23557775/72b250ec-0046-11e7-9631-daf2890994c8.jpg)


**2) Interrupting at 2000 and re-starting by the produced backup weights:**

![2001](https://cloud.githubusercontent.com/assets/18719591/23557801/89698940-0046-11e7-8354-e1484bfb052d.jpg)

**Comment by AlexeyAB** [comment]
Interestingly. I do not know why this is happening. Is it reproduced repeatedly?

There might be several reasons, but none of them explains this behavior:

1. `avg_loss = avg_loss*.9 + loss*.1;` - accumulates errors, but it quickly updated: https://github.com/AlexeyAB/darknet/blob/b3a3e92e8a482cf1c49322561077e8f3be54d619/src/detector.c#L136

2. In earlier versions, this could be due to the malfunctioning of the random number generator, but this bug fixed for a long time ago: https://github.com/AlexeyAB/darknet/commit/b831db5e58ad37c8f4adf39de7fb5db204209a1c#diff-2ceac7e68fdac00b370188285ab286f7R719

**Comment by VanitarNordic** [comment]
I'm not quite sure if this is a bug or not. maybe the application does not start from where it left necessarily. I mean it depends on what batches are ready for processing at the run-time maybe.
I restarted it several times, sometimes it start from the almost same avg error and sometimes no. That's the reason why I say it "could" be normal.

**Comment by AlexeyAB** [comment]
I think it is not worth paying attention on it. Because images for training are selected randomly, and if selected the same images that are already has been selected earlier for training, then the error will be very low - because network already trained for it.

I started to training a few times since 9000 iterations and `avg_value` **varies greatly** depending on how Yolo randomly selects images. **avg_value = from 0.02 to 0.15**

**Comment by VanitarNordic** [comment]
Yes, Thank you. This is the end of this story. 

Just as the reminder. Have you checked the yolo_mark issue? also background images
